codegen/README.md

```diff
@@ -0,0 +1,50 @@
+# Codegen
+
+This crate provides code-generation for the stm32f3xx-hal. It reads information
+from an [STM32CubeMX](https://www.st.com/en/development-tools/stm32cubemx.html)
+database and uses that to output code that can directly be included into the
+source code of the stm32f3xx-hal crate.
+
+For more information on how the STM32CubeMX database is structured, check out
+the README in the [cube-parse](https://github.com/dbrgn/cube-parse) repository.
+
+Because by default cargo tries to use the `thumbv7em-none-eabihf` target, due
+to what's specified in the `.cargo/config`, you need to manually specify your
+host's target when building `codegen`, e.g.:
+
+```
+$ cargo run --target x86_64-unknown-linux-gnu -- help
+```
+
+`codgen` can generate the following code:
+
+- GPIO mappings
+
+## GPIO mappings
+
+Running `codegen`'s `gpio` subcommand generates the `gpio!` macro
+invocations at the end of `src/gpio.rs`. Re-generating those macro-invocations
+is simply a matter of deleting the old ones and then executing:
+
+```
+$ cargo run --target $host_target -- gpio $cubemx_db_path >> ../src/gpio.rs
+```
+
+`$cubemx_db_path` must be the path to the `db/mcu` directory under an
+STM32CubeMX installation. With a default Linux install, this would be
+`/opt/stm32cubemx/db/mcu`.
+
+The generated `gpio!` invocations are gated by features whose names are derived
+from the respective GPIO internal peripheral (IP) version:
```

Comment:
Yes, codegen pulls all its information from a CubeMX DB, that's mentioned in the beginning of the README. Are you saying I should mention that explicitly here again?

And yes, at some point it would be nice to auto-generate also the features in the Cargo.toml. Right now it doesn't really work because we are using custom MCU features like "stm32f303xc" which don't fit the MCU (sub-)families known to CubeMX. We could solve this by moving to more fine-grained features (like the L0 HAL has), but that would be another large change and for now hand-writing the feature definitions isn't much effort at all.

Reply:
> Yes, codegen pulls all its information from a CubeMX DB, that's mentioned in the beginning of the README. Are you saying I should mention that explicitly here again?

To me as a reader it is unclear where the gpio ip version is taken from. As we are adding those manually to the Cargo.toml, I feel like it should be mentioned here again.
 
> And yes, at some point it would be nice to auto-generate also the features in the Cargo.toml. Right now it doesn't really work because we are using custom MCU features like "stm32f303xc" which don't fit the MCU (sub-)families known to CubeMX. We could solve this by moving to more fine-grained features (like the L0 HAL has), but that would be another large change and for now hand-writing the feature definitions isn't much effort at all.

totally agree.